sumcheck/src/prover/high_degree.rs

```diff
@@ -0,0 +1,488 @@
+/// we can write
+///
+/// $$
+///     f_i(X_i, x_{i + 1}, \cdots, x_{\nu - 1}) =
+///        (1 - X_i) . f_i(0, x_{i + 1}, \cdots, x_{\nu - 1}) +
+///        X_i . f_i(1, x_{i + 1}, \cdots, x_{\nu - 1})
+/// $$
+///
+/// Note that we omitted writing the folding randomness for readability.
+/// Since the evaluation domain is $\{0, 1, ... , d_max\}$, we can compute the evaluations based on
+/// the previous one using only additions. This is the purpose of `deltas`, to hold the increments
+/// added to each multi-linear to compute the evaluation at the next point, and `evals_x` to hold
+/// the current evaluation at $x$ in $\{2, ... , d_max\}$.
+fn sumcheck_round<E: FieldElement>(
+    eq_mu: &[E],
+    evaluator: &impl LogUpGkrEvaluator<BaseField = <E as FieldElement>::BaseField>,
+    eq_ml: &MultiLinearPoly<E>,
+    mls: &[MultiLinearPoly<E>],
+    log_up_randomness: &[E],
+    r_sum_check: E,
+) -> CompressedUnivariatePolyEvals<E> {
+    let num_ml = mls.len();
+    let num_vars = mls[0].num_variables();
+    let num_rounds = num_vars - 1;
+
+    #[cfg(not(feature = "concurrent"))]
+    let evaluations = {
+        let mut evals_one = vec![E::ZERO; num_ml];
+        let mut evals_zero = vec![E::ZERO; num_ml];
+        let mut evals_x = vec![E::ZERO; num_ml];
+        let mut eq_x = E::ZERO;
+
+        let mut deltas = vec![E::ZERO; num_ml];
+        let mut eq_delta = E::ZERO;
+
+        let mut numerators = vec![E::ZERO; evaluator.get_num_fractions()];
+        let mut denominators = vec![E::ZERO; evaluator.get_num_fractions()];
+        (0..1 << num_rounds)
+            .map(|i| {
+                let mut total_evals = vec![E::ZERO; evaluator.max_degree()];
+
+                for (j, ml) in mls.iter().enumerate() {
+                    evals_zero[j] = ml.evaluations()[2 * i];
+                    evals_one[j] = ml.evaluations()[2 * i + 1];
+                }
+
+                let eq_at_zero = eq_ml.evaluations()[2 * i];
+                let eq_at_one = eq_ml.evaluations()[2 * i + 1];
+
+                // compute the evaluation at 1
+                evaluator.evaluate_query(
+                    &evals_one,
+                    log_up_randomness,
+                    &mut numerators,
+                    &mut denominators,
+                );
+                total_evals[0] = evaluate_composition_poly(
+                    eq_mu,
+                    &numerators,
+                    &denominators,
+                    eq_at_one,
+                    r_sum_check,
+                );
+
+                // compute the evaluations at 2, ..., d_max points
+                evals_zero
+                    .iter()
+                    .zip(evals_one.iter().zip(deltas.iter_mut().zip(evals_x.iter_mut())))
+                    .for_each(|(a0, (a1, (delta, evx)))| {
+                        *delta = *a1 - *a0;
+                        *evx = *a1;
+                    });
```

Comment:
Is this equivalent to:
```rust
for i in 0..evaluator.max_degree() {
    deltas[i] = evals_one[i] - evals_zero[i];
    evals_x[i] = evals_one[i];
}
```
If so, I'd probably use the above as I think it is much more readable.

Reply:
Simplified it using your proposition